recipes/jenkins.rb

```diff
@@ -31,6 +31,53 @@
 apt_update
 
 package 'openjdk-8-jdk-headless'
+# Jenkins downgrade protection
+#
+# The Jenkins package has transitioned to using systemd units instead of
+# sysvinit style init scripts. In order to maintain this cookbook the last
+# version of Jenkins which is packaged with a sysvinit script is pinned.
+# However Jenkins cannot be safely downgraded and some configurations of
+# Jenkins will continue to function once initially configured even after the
+# systemd switch.
+# To protect users from unintentional downgrades we're going to do something really ugly here.
+package 'jenkins' do
+  action :lock
+  only_if { -> do
+    jenkins_info = `dpkg -s jenkins`
+    if $?.exitstatus != 0
+       return false
+    end
+
+    version_line = jenkins_info.lines.select{|line| line =~ /^Version: /}.first
+    if version_line.nil?
+      return false
```

Comment:
I'm not actually sure that this code is reachable, I can't think of a situation where the command above would exit successfully but there wouldn't be any version info from it.

I may decide to take this out or turn it into a hard error because it's "wrong"

Reply:
Eh, you're parsing output from a command line tool. I think it's totally reasonable to have a protection like this.

Not that you should support the scenario, but I could see something like localization resulting in `nil` here.

Reply:
> I could see something like localization resulting in `nil` here.

Yeah, I think you're right and on consideration if we get here we should make the user stop and figure out how and what to do about it. 9fb0a780ad178cffda22568ac67b4d706be9c1bc